File: shared/cost_functions.py
import math
import copy 
from functools import reduce
from shared.models.task import Task

# not used anymore 
def get_ready(task_set, cycle, ready_list):
    for task in task_set:
        if cycle % task.period == 0:
            new_job = copy.deepcopy(task)
            new_job.deadline = new_job.deadline + cycle
            new_job.release_time = cycle
            ready_list.append(new_job)
    
    # sort on deadline 
    ready_list = sorted(ready_list, key=lambda t: t.deadline) # https://www.geeksforgeeks.org/sorting-objects-of-user-defined-class-in-python/ 
    
    return ready_list

# processor demand. see section 4.6 of hard real time 
def processor_demand_criterion(t1, t2, task_set):

    # phase or capital phi is 0 for all task in our case
    # for some task demand = duration*number_of_instances_interval 
    demand = sum([max(0, (t2 + task.period - task.deadline - t1) / task.period) * task.duration for task in task_set])
    
    # task set is schedulable if processor demand does not exceed available time  
    return demand <= (t2 - t1)

# release new task instance
def create_job(cycle, task): # create new instance instead of deepcopy. Tried using list instead of task object. doesnt seem better and less readble 
    new_job = Task(task.name, task.duration, task.period, task.type, task.priority, task.deadline + cycle, et_subset=task.et_subset)
    #new_job.deadline = new_job.deadline + cycle
    new_job.release_time = cycle

    return new_job 

# microticks to get to next point at which some task will be relased
def steps_to_next_release(cycle, task_set):
    return min([task.period - (cycle % task.period) for task in task_set])
        
# ts is set of TT tasks
def edf(ts):
    periods = [t.period for t in ts]
    T = math.lcm(*periods) # least common multiple of TT task periods
    s = [] # schedule will be hyperperiod long. 12 000 microticks == 120 000 microsecs == 120 ms
    ready_list = [] 
    wcrts = {} # worst case response times
    is_schedulable = True
    # total amount of computation time needed exceeding available time -> not schedulable
    if not processor_demand_criterion(0, T, ts): 
        #return [], -1 
        is_schedulable = False

    # return same thing in all failed cases, just empty list fx!! 
    t = 0
    while t < T:
        for task in ready_list:
            if task.duration > 0 and task.deadline <= t:
                #return [], -1 # just return 1 here maybe lol?
                is_schedulable = False

        # release task at time t
        ready_list = ready_list + [create_job(t, task) for task in ts if t % task.period == 0]     

        if ready_list == []:
            # increase period to next task's release
            steps_to_skip = steps_to_next_release(t, ts)
            t += steps_to_skip
            
            # append steps_to_skip amount of IDLEs to schedule
            s.extend(["IDLE"] * steps_to_skip)
            continue
        else:
            # EDF get next job to execute  
            ed_job = min(ready_list, key=lambda x: x.deadline)
            s.append(ed_job.name)
            ed_job.duration = ed_job.duration - 1

            # 31 october addition to assignment
            if ed_job.duration == 0 and ed_job.deadline >= t:
                response_time = t - ed_job.release_time 
                
                if ed_job.name not in wcrts or response_time >= wcrts[ed_job.name]: 
                    wcrts[ed_job.name] = response_time 
                 
                ready_list.remove(ed_job)

        # increment cycle counter
        t += 1

    # not feasible if ready list is not empty after hyperperiod
    if ready_list != []:
        #return [], -1
        is_schedulable = False

    #print("in EDF wcrts is: ", wcrts)
    return s, wcrts, is_schedulable

# utility function 
def unpack(task):
    return (task.priority, task.duration, task.period, task.deadline)

#Schedulability of ET tasks under a given polling task
def calculate_schedulabiltiy(polling_server):
    et_tasks = polling_server.et_subset
    Tp = polling_server.period
    Dp = polling_server.deadline
    Cp = polling_server.duration
    result_dict = {}
    is_schedulable = True
    #compute delta and alpha accordingly to [2]
    Delta = Tp + Dp - 2*Cp
    #Original -> Using lower bound: alpha = Cp/Tp (from [1]^2)
    #We use the so-called linear supply lower bound function lslbf (t) (c.f. [9]) with α = Cp/Tp and ∆ = Tp + Dp − 2*Cp, as lslbf (t) = max{0, (t − ∆)*α}.
    #Ext3: Use the actual server supply function called A_s(t) which can be determined from the schedule table
    #   This will be done by remove alpha, and thereby saying the supply at e.g tick 2 is 2 - that way it will remove the pessimism
    alpha = Cp/Tp

    #hyperperiod is lcm of all task periods in T_ET (all values must be from the chosen subset of ET tasks from the .csv)
    periods = [t.period for t in et_tasks]
     
    hyperperiod = math.lcm(*periods)
    
    for et_task in et_tasks:
        (pi, Ci, Ti, Di) = unpack(et_task)
        t = 0
        #initialize the response time of ti (task period) to a value exceeding the deadline
        response_time = Di + 1

        #remember we are dealing with constrained deadline tasks for the AdvPoll, hence, in the worst case arrival pattern, the intersection must lie within the hyperperiod if the task is schedulable

        while t <= hyperperiod:
            #the supply at time t ([1])
            supply = alpha*(t-Delta)
            #EXT3:
            #supply = t-Delta

            #compute the maximum demand at time t according to Eq. 2
            demand = 0
            for tj in et_tasks:
                (pj, Cj, Tj, Dj) = unpack(tj)

                if pj >= pi:
                    demand = demand + math.ceil(t/Tj)*Cj
            
            #According to lemma 1 of [1], we are searching for the earliest time, when the supply exceeds the demand
            if supply >= demand:
                response_time = t
                result_dict[et_task.name] = (response_time, et_task.deadline) # if actually greater than deadline, set to false later
                break
            
            t = t + 1
        
        if response_time > Di: 
            is_schedulable = False # TODO maybe just return here ... and penalize with 1 
            result_dict[et_task.name] = (response_time, et_task.deadline)
            return is_schedulable, result_dict 
        
    return is_schedulable, result_dict # contains wcrtbool indicating schedulability and deadline for each et

# sum of average worst case response time for tt tasks and et tasks. penality for not schedulable
# also returns the schedule and boolean indicating schedulability 
def cost_f(task_set):
    polling_servers = [ps for ps in task_set if ps.et_subset != None] # get set of polling servers from task set
    l = [calculate_schedulabiltiy(ps) for ps in polling_servers] # check schedulability for each polling server
    
    wcrts_et = 0 # use worst case response time for cost metric
    is_schedulable = True # if some ps is not schedulable at some penalty to cost 
    
    # cost contribution of et tasks is the average of worst case response times
    # si costs for et is: (sum(wcrt_i/deadline_i) / len(et_subset)) /len(polling_servers)
    for element in l: 
        is_schedulable, wcrts = element
        
        wcrts_et += sum([wcrts[key][0] for key in wcrts]) / len(wcrts)
        
        if not is_schedulable:
            wcrts_et = wcrts_et + 0.2*wcrts_et # add penalty
         
    # do not divide by zero. 
    if l != []:
        wcrts_et *= 1/len(l) 
     
    # apply earliest deadline first 
    s, wcrts, is_schedulable = edf(task_set) 
    
    # handle case where some task isn't even run also by check if in dict
    sum_wcrts_tt = sum([wcrts[task.name] if task.name in wcrts else task.deadline + 100 for task in task_set ]) / len(task_set)
    if not is_schedulable: # penalize. do this to avoid ending up in a "false" minimum
        sum_wcrts_tt = sum_wcrts_tt + 0.2*sum_wcrts_tt 

    #print("cost et: ", wcrts_et, " cost tt: ", sum_wcrts_tt) 
    sum_avg_wcrts = (sum_wcrts_tt + wcrts_et)
 
    return s, sum_avg_wcrts, is_schedulable

File: genetic_algorithm/src/cost_functions.cpp
#include "cost_functions.h"
#include <iostream>

/*

    TODO rand thing ....

*/


double cost_function(std::vector<Task> *task_set) {
    std::list<Task> polling_servers; 
    std::map<std::string, double> wcrts_tt; // doubles because we want to take an average
    std::map<std::string, double> wcrts_et; // reassigned when iterating over polling servers
    std::tuple<bool, std::map<std::string, double>, std::vector<std::string>> is_schedulable_cost_tt;
    std::tuple<bool, std::map<std::string, double>> is_schedulable_cost_et; 
    double cost_tt = 0.0, cost_et = 0.0, cost = 0.0, wcrts_et_sum = 0.0, wcrts_tt_sum = 0.0;
    bool is_schedulable;

    for (auto it : *task_set) { 
        if (it.et_subset != NULL) { polling_servers.push_back(it); };        
    }

    // get cost contribution of each polling server 
    for(auto it : polling_servers) {
        is_schedulable_cost_et = is_polling_server_schedulable(&it);
        is_schedulable = std::get<0>(is_schedulable_cost_et);
        wcrts_et_sum = 0;
        wcrts_et = std::get<1>(is_schedulable_cost_et);
        
        // if polling server is schedulable sum wcrts if not sum deadline + a penalty for each
        for (auto et_task : *it.et_subset) {
            if (is_schedulable) { wcrts_et_sum += wcrts_et[et_task.name]; }
            else { wcrts_et_sum += 2*et_task.deadline; }
        }

        cost_et += wcrts_et_sum / it.et_subset->size();
    }

    // do not divide by zero
    cost_et = polling_servers.size() > 0 ? cost_et / polling_servers.size() : 0; 
        
    is_schedulable_cost_tt = edf(task_set);
    is_schedulable = std::get<0>(is_schedulable_cost_tt);
    wcrts_tt = std::get<1>(is_schedulable_cost_tt);

    // same approach as for ets in a ps. wcrt or deadline + penalty
    for(auto it : *task_set) {
        if(is_schedulable) { wcrts_tt_sum += wcrts_tt[it.name]; }
        else {wcrts_tt_sum += 2*it.deadline; }
    }

    cost_tt = wcrts_tt_sum / task_set->size();
     
    cost = cost_tt + cost_et;
    //std::cout << "cost function cost is: " << cost << " et: " << cost_et << " tt: " << cost_tt << std::endl;
    return cost; 
}

// also returns is_schedulable
std::tuple<double, bool>  cost_function1(std::vector<Task> *task_set) {
    std::list<Task> polling_servers; 
    std::map<std::string, double> wcrts_tt; // doubles because we want to take an average
    std::map<std::string, double> wcrts_et; // reassigned when iterating over polling servers
    std::tuple<bool, std::map<std::string, double>, std::vector<std::string>> is_schedulable_cost_tt;
    std::tuple<bool, std::map<std::string, double>> is_schedulable_cost_et; 
    double cost_tt = 0.0, cost_et = 0.0, cost = 0.0, wcrts_et_sum = 0.0, wcrts_tt_sum = 0.0;
    bool is_schedulable; // iteration
    bool is_schedulable_et=true, is_schedulable_tt;
    int num_et_tasks = 0;

    for (auto it : *task_set) { 
        if (it.et_subset != NULL) { polling_servers.push_back(it); };        
    }

    // get cost contribution of each polling server 
    for(auto it : polling_servers) {
        is_schedulable_cost_et = is_polling_server_schedulable(&it);
        is_schedulable = std::get<0>(is_schedulable_cost_et);
        is_schedulable_et = is_schedulable_et && is_schedulable;
        wcrts_et_sum = 0;
        wcrts_et = std::get<1>(is_schedulable_cost_et);
        
        // if polling server is schedulable sum wcrts if not sum deadline + a penalty for each
        for (auto et_task : *it.et_subset) {
            num_et_tasks = num_et_tasks + 1;
            if (is_schedulable) { wcrts_et_sum += wcrts_et[et_task.name]; }
            else { wcrts_et_sum += 2*et_task.deadline; }
        }

        cost_et += wcrts_et_sum; /// it.et_subset->size();
    }

    // do not divide by zero
    //cost_et = polling_servers.size() > 0 ? cost_et / polling_servers.size() : 0; 
        
    is_schedulable_cost_tt = edf(task_set);
    is_schedulable_tt = std::get<0>(is_schedulable_cost_tt);
    wcrts_tt = std::get<1>(is_schedulable_cost_tt);

    // same approach as for ets in a ps. wcrt or deadline + penalty
    for(auto it : *task_set) {
        if(is_schedulable_tt) { wcrts_tt_sum += wcrts_tt[it.name]; }
        else {wcrts_tt_sum += 2*it.deadline; }
    }

    cost_tt = wcrts_tt_sum;// / task_set->size();
     
    // (sum wcrt tt + sum wcrt et) 
    cost = (cost_tt + cost_et) / (task_set->size() + num_et_tasks);
    //std::cout << "cost function cost is: " << cost << " et: " << cost_et << " tt: " << cost_tt << std::endl;
    return std::tuple<double, bool>(cost, is_schedulable_et && is_schedulable_tt); 
}

// release jobs 
void get_ready_list(std::vector<Task> *taskSet, std::list<Task> *readyList, int cycle) {  
    for(auto task : *taskSet) {
        if (cycle % task.period == 0) {
            // release new instance of task with deadline = Di + t and release time = t
            readyList->push_back(Task(task.name, task.duration, task.period, task.type, task.priority, task.deadline + cycle, cycle));
        }
    }    
}

/* deadline used here is relative deadline. section 4.6 hard real time 
   only for t1 = 0 not general case. also assumes that Di <= Ti
*/
int processor_demand(int t2, std::vector<Task> *periodic_tasks) {
    int n = periodic_tasks->size(); 
    int processor_demand=0; 
    double eta_i, period_i, deadline_i;

    // demand for some task is number of times scheduled * duration
    for(int i=0; i<n; i=i+1) {
        period_i = (double) periodic_tasks->at(i).period;
        deadline_i = (double) periodic_tasks->at(i).deadline; // line would be really long if we did this in line below
        eta_i = (t2 + period_i - deadline_i) /  period_i; // we could do without 2nd term but more general here

        processor_demand +=  ((int) eta_i) * periodic_tasks->at(i).duration;
    }

    return processor_demand;
}

//https://www.wolframalpha.com/input?i=lcm%28a%2Cb%2Cc%29+%3D%3D+lcm%28lcm%28a%2Cb%29%2C+c%29
//https://en.cppreference.com/w/cpp/numeric/lcm
//https://www.geeksforgeeks.org/lcm-of-given-array-elements/ 
std::tuple<bool, std::map<std::string, double>, std::vector<std::string>> edf(std::vector<Task> *tt_tasks) { 
    int hyperperiod = tt_tasks->front().period;
    
    // https://stackoverflow.com/questions/4210470/looping-on-c-iterators-starting-with-second-or-nth-item 
    for(auto it = std::next(tt_tasks->begin()); it != tt_tasks->end(); ++it) {
        hyperperiod = std::lcm(hyperperiod, it->period);
    }
    
    //std::cout << "hyperperiod is: " << hyperperiod << std::endl;
    std::map<std::string, double> wcrts;
    std::vector<std::string> schedule;

    if (processor_demand(hyperperiod, tt_tasks) > hyperperiod) {
        return std::tuple(false, wcrts, schedule);
    }  

    std::list<Task> ready_list; 
    int t = 0;

    // TODO: skip if none ready
    while (t < hyperperiod) {
        
        // check for deadline miss
        for (auto job : ready_list) {
            if (job.duration > 0 and job.deadline <= t) {
                return std::tuple(false, wcrts, schedule); 
            }
        }

        // release new task instances 
        get_ready_list(tt_tasks, &ready_list, t);

        // execute task instance with earliest deadline, if there are any ready tasks
        if (ready_list.size() == 0) {
            schedule.push_back("IDLE"); // no tasks instantiated, nothing to do 
        } else {

            // get task instance with earliest deadline https://stackoverflow.com/questions/26766136/how-to-get-min-or-max-element-in-a-vector-of-objects-in-c-based-on-some-field 
            auto ed_task = std::min_element(ready_list.begin(), ready_list.end(), Task::ByDeadline());  
            schedule.push_back(ed_task->name);
            ed_task->duration -= 1;

            // check if current response time larger than than current maximum
            if (ed_task->duration == 0 && ed_task->deadline >= t) {
                ready_list.erase(ed_task); // remove from ready list 
                if (!wcrts.contains(ed_task->name) || t - ed_task->release_time >= wcrts[ed_task->name]) {
                    wcrts[ed_task->name] = t - ed_task->release_time;
                }
            } else if(ed_task->duration == 0 && ed_task->deadline < t) { // this check should be redundant but to be safe. should be caught above if this is the case 
                return std::tuple(false, wcrts, schedule);  
            }
        }

        t = t + 1;
    }

    return std::tuple(true, wcrts, schedule); 
}

// determine if some polling server with a given budget, period, deadline and set of et tasks is schedulable
std::tuple<bool, std::map<std::string, double>> is_polling_server_schedulable(Task* polling_server) {
    
    int budget = polling_server->duration;
    int period = polling_server->period;
    int deadline = polling_server->deadline;
    std::vector<Task>* et_set = polling_server->et_subset;  
    std::map<std::string, double> response_times;

    // does not matter what we return in this case 
    if (et_set == NULL) {
        return std::tuple(true, response_times);
    }

    int delta = period + deadline - 2*budget;
    double alpha = double(budget) / double(period);
    double supply, demand, cur_response_time; 
    
    

    int hyperperiod = et_set->front().period;
    
    // https://stackoverflow.com/questions/4210470/looping-on-c-iterators-starting-with-second-or-nth-item
    for(auto it = std::next(et_set->begin()); it != et_set->end(); ++it) {
        hyperperiod = std::lcm(hyperperiod, it->period);
    }


    for(auto et_it : *et_set) {
        // initialize response time of current et task to a value exceeding its deadline
        cur_response_time = et_it.deadline + 1;
        
        for(int t = 0; t <= hyperperiod; t = t + 1) {
            supply = alpha*(double(t) - double(delta));
            demand = 0;

            for(auto et_it_j : *et_set) {

                // compute maximum demand at time t
                if (et_it_j.priority >= et_it.priority) { 
                    demand = demand + std::ceil(double(t) / double(et_it_j.period)) * double( et_it_j.duration);
                }
            } // end of traversal of other ets

            if (supply >= demand) { 
                cur_response_time = t;
                response_times[et_it.name] = cur_response_time; 
                break; 
            }
        } // end of inner for

        // do not return here ! 
        if (cur_response_time >= et_it.deadline) { 
            return std::tuple<bool, std::map<std::string, double>>(false, response_times);
        }    

    } // end of outer for

    return std::tuple<bool, std::map<std::string, double>>(true, response_times);

}
